src/dataflow/src/sink/kafka.rs

```diff
@@ -1009,6 +1036,11 @@ where
                 return true;
             }
 
+            // N.B. Given the `flush` call above, I don't think we should ever end up in this
+            // situation but let's keep the metrics / logging here so we can verify this in real
+            // world use cases.
+            let in_flight = s.producer.in_flight_count();
+            s.metrics.messages_in_flight.set(in_flight as u64);
             if in_flight > 0 {
                 // We still have messages that need to be flushed out to Kafka
```

Comment:
If this is an impossible situation can we `assert` it instead?

Reply:
I'm not sure that it's impossible -- I just think it might be.  I think it's possible that `maybe_emit_progress` above could leave unflushed records?

Since we already have the metrics set up, I'd like to just leave it in until we can look at them more closely and see if it is in fact always zero.

The downside of being wrong here is pretty high (sinks are unusable)

Reply:
I looked a bit into and it is indeed impossible:

https://docs.confluent.io/3.1.1/clients/librdkafka/rdkafka_8h.html#aaff06c4372bce917c17f3c1a5d8b205d

> Wait until all outstanding produce requests, et.al, are completed. This should typically be done prior to destroying a producer instance to make sure all queued and in-flight produce requests are completed before terminating. 

and this SO link: https://stackoverflow.com/questions/40144502/librdkafka-what-is-the-purpose-of-rd-kafka-outq-len-when-there-is-rd-kafka-flus

Reply:
Yeah -- I've read both of those.  I meant more "impossible" w.r.t. our code.  But I'll just go ahead and add a flush at the end where this currently is and remove any `in_flight_count` instrumenting